{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GermlineMasterWorkflow.wdl output check development notes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "x_dir = '/Users/yo/zzIForge/fullyJan10/'\n",
    "if not os.path.isdir(x_dir):\n",
    "    print('directory not found\\n', x_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rc_codes_df(x_dir):\n",
    "    \"\"\" Usage: return_codes_dataframe = get_rc_codes_df(x_dir)\n",
    "    args:\n",
    "        x_dir:         the directory with the \"call_...\" subdirectories (else you get nothing)\n",
    "    returns:\n",
    "        rc_df:         pandas dataframe with the return codes and size of various return files\n",
    "    \"\"\"\n",
    "    good_return_codes_list = ['0', '0\\n']\n",
    "    \n",
    "    call_dirs = os.listdir(x_dir)\n",
    "    call_dir_list = []\n",
    "    call_dir_count = 0\n",
    "    for cd in call_dirs:\n",
    "        if os.path.isdir(os.path.join(cd,x_dir)) and cd[0:4] == 'call':\n",
    "            call_dir_count += 1\n",
    "            call_dir_list.append(cd)\n",
    "    \n",
    "    cols_list = ['rc', 'bam', 'bam.bai']\n",
    "    #rc_data = np.zeros((call_dir_count,len(cols_list))) - 1\n",
    "    rc_df = pd.DataFrame(index=call_dir_list,columns=cols_list).fillna('unk')\n",
    "    root_trim_str, _ = os.path.split(x_dir)\n",
    "    \n",
    "    for dir_name, dir_list, files_list in os.walk(x_dir):\n",
    "        for filename in files_list:\n",
    "            full_filename = os.path.join(dir_name, filename)\n",
    "            if filename in cols_list:\n",
    "                if filename == 'rc':\n",
    "\n",
    "                    with open(full_filename, 'r') as fh:\n",
    "                        lines = fh.readlines()\n",
    "\n",
    "                    if lines[0] in good_return_codes_list:\n",
    "                        for call_dir in call_dir_list:\n",
    "                            if call_dir in dir_name:\n",
    "                                top_dir = dir_name.replace(root_trim_str, '..')\n",
    "                                rc_df.loc[call_dir, 'rc'] = str(lines[0]).strip()\n",
    "\n",
    "            else:\n",
    "                for call_dir in call_dir_list:\n",
    "                    if call_dir in dir_name:\n",
    "                        fname, fext = os.path.splitext(filename)\n",
    "                        some_stuff = os.stat(full_filename)\n",
    "                        if fext[1:] == 'bam':\n",
    "                            rc_df.loc[call_dir, 'bam'] = str(some_stuff.st_size)\n",
    "                        elif fext[1:] == 'bai':\n",
    "                            rc_df.loc[call_dir, 'bam.bai'] = str(some_stuff.st_size)\n",
    "                                \n",
    "    return rc_df                 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "somedf = get_rc_codes_df(x_dir)\n",
    "somedf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## wdl sequence of calls:\n",
    "    * BashPreamble: src/shell/shell_preamble.sh (every task starts with it)\n",
    "    * most scripts use (using shared_functions.sh)\n",
    "1) GermlineMasterWorkflow.wdl (iffie) calls TestTasks/Runtrim_sequences.wdl calls Tasks/trim_sequences.wdl runs:\n",
    "    */bin/bash trim_sequences.sh (PairedEnd, SampleName, Adapters, CutAdapt, CutAdaptThreads, TrimEnvProfile)\n",
    "        * TrimSeqScript = src/shell/trim_sequences.sh (uses shared_functions.sh)\n",
    "        * BashSharedFunctions = src/shell/shared_functions.sh\n",
    "    * parameter settings\n",
    "        * TrimEnvProfile = Config/TrimEnvProfile.file\n",
    "        * CutAdapt = /usr/local/apps/bioapps/python/Python-3.6.1/bin\n",
    "        * PairedEnd = true or false\n",
    "        * SampleName = WGS_chr1_5X_E0.005_chunk1\n",
    "        * Adapters = /Inputs/TruSeqAdaptors.fasta\n",
    "        * CutAdaptThreads = 8\n",
    "        * DebugMode = DNE\n",
    "    * Puts Out: glob(\"*.fastq.gz\") somewhere in hyper-cyber space\n",
    "    \n",
    "3) GermlineMasterWorkflow.wdl calls TestTasks/Runalignment.wdl (diffie) calls Tasks/alignment.wdl runs:\n",
    "    * AlignmentScript: src/shell/alignment.sh \n",
    "        * calls: Sentieon /usr/local/apps/bioapps/sentieon/sentieon-genomics-201808.01\n",
    "    * Warns: cpu, s_vmem, h_vmem\n",
    "    * Puts Out:\n",
    "        * OutputBams: SampleName.aligned.sorted.bam\n",
    "        * OutputBais: SampleName.aligned.sorted.bam.bai\n",
    "    \n",
    "4) GermlineMasterWorkflow.wdl calls Tasks/merge_aligned_bam.wdl runs:\n",
    "    * MergeBamScript: src/shell/merge_bams.sh (using shared_functions.sh)\n",
    "    * with Config/MergeBamEnvProfile.file\n",
    "    * Warns: cpu, s_vmem, h_vmem\n",
    "    * Puts Out:\n",
    "        * OutputBams: SampleName.aligned.sorted.merged.bam\n",
    "        * OutputBais: SampleName.aligned.sorted.merged.bam.bai\n",
    "        \n",
    "5) GermlineMasterWorkflow.wdl calls Tasks/dedup.wdl runs:\n",
    "    * DedupScript: src/shell/dedup.sh (using shared_functions.sh)\n",
    "    * with Config/DedupEnvProfile.file\n",
    "    * Warns: cpu, s_vmem, h_vmem\n",
    "    * Puts Out:\n",
    "        * OutputBams: SampleName.aligned.sorted.deduped.bam\n",
    "        * OutputBais: SampleName.aligned.sorted.deduped.bam.bai\n",
    "        \n",
    "6) GermlineMasterWorkflow.wdl calls Tasks/deliver_alignment.wdl runs:\n",
    "    * DeliveryAlignment_Script: src/shell/deliver_alignment.sh (using shared_functions.sh)\n",
    "    * using: GermlineMasterWorkflow.FilledIn.json\n",
    "    * no output but: dowop Delivery/Alignment\n",
    "    \n",
    "7) GermlineMasterWorkflow.wdl calls Tasks/realignment.wdl runs:\n",
    "    * RealignmentScript: src/shell/realignment.sh (using shared_functions.sh)\n",
    "    * using: Reference/Mills_and_1000G_gold_standard.indels.hg38.vcf\n",
    "    * RealignEnvProfile: Config/RealignEnvProfile.file\n",
    "    * Warns: cpu, s_vmem, h_vmem\n",
    "    * Puts Out:\n",
    "        * OutputBams: SampleName.aligned.sorted.deduped.realigned.bam\n",
    "        * OutputBais: SampleName.aligned.sorted.deduped.realigned.bam.bai\n",
    "        \n",
    "8) GermlineMasterWorkflow.wdl calls Tasks/bqsr.wdl runs:\n",
    "    * BqsrScript: src/shell/bqsr.sh (\n",
    "    * using: \n",
    "        * Reference/Mills_and_1000G_gold_standard.indels.hg38.vcf\n",
    "        * Reference/dbsnp_138.hg38.vcf\n",
    "        * Config/BqsrEnvProfile.file\n",
    "    * Warns: cpu, s_vmem, h_vmem\n",
    "    * Puts Out:\n",
    "        * RecalTable: SampleName.recal_data.table\n",
    "        \n",
    "9) GermlineMasterWorkflow.wdl calls Tasks/haplotyper.wdl runs:\n",
    "    * HaplotyperScript: src/shell/haplotyper.sh (using shared_functions.sh)\n",
    "    * using:\n",
    "        * HaplotyperExtraOptionsString: (magic string)\n",
    "        * Config/HaplotyperEnvProfile.file\n",
    "        * Config/HaplotyperVCFSourceField.file\n",
    "    * Warns: cpu, s_vmem, h_vmem\n",
    "    * Puts Out:\n",
    "        * OutputVcf: SampleName.vcf\n",
    "        * OutputVcfIdx: SampleName.vcf.idx\"\n",
    "        \n",
    "10) GermlineMasterWorkflow.wdl calls Tasks/vqsr.wdl runs:\n",
    "    * VqsrScript: src/shell/vqsr.sh \n",
    "    * using magic strings:\n",
    "        * VqsrSnpResourceString\n",
    "        * VqsrIndelResourceString\n",
    "        * AnnotateText\n",
    "    * using:\n",
    "        * InputVcf && InputVcfIdx (input output from) VCF from Haplotyper \n",
    "        * Config/VqsrEnvProfile.file\n",
    "    * Warns: cpu, s_vmem, h_vmem\n",
    "    * Puts Out:\n",
    "        * OutputVcf: SampleName.SNP.recaled.vcf\n",
    "        * OutputVcfIdx: SampleName.SNP.recaled.vcf.idx\n",
    "        \n",
    "10) GermlineMasterWorkflow.wdl calls Tasks/deliver_HaplotyperVC.wdl runs:\n",
    "    * DeliveryHaplotyperVC_Script: src/shell/deliver_haplotyperVC.sh (using shared_functions.sh)\n",
    "    * to dakine: Delivery/HaplotyperVC\n",
    "    * using: GermlineMasterWorkflow.FilledIn.json    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7 GermlineMasterWF.DAB.WorkflowJson\n",
      "\t/projects/bioinformatics/DEL/Jsons/GermlineMasterWorkflow.FilledIn.json\n",
      "\n",
      "14 GermlineMasterWF.DHVC.WorkflowJson\n",
      "\t/projects/bioinformatics/DEL/Jsons/GermlineMasterWorkflow.FilledIn.json\n",
      "\n"
     ]
    }
   ],
   "source": [
    "json_dir = '/Users/yo/zzIForge/Jsons'\n",
    "json_farameters_file = 'GermlineMasterWorkflow.FilledIn.json'\n",
    "import knpackage.toolbox as kn\n",
    "json_parameters = kn.get_run_parameters(json_dir, json_farameters_file)\n",
    "dakine = 'WorkflowJson'\n",
    "lineno = 0\n",
    "for k, v in json_parameters.items():\n",
    "    if dakine in k:\n",
    "        print('%i %30s\\n\\t%s\\n'%(lineno, k, v))\n",
    "    lineno += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
