{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# XxxWorkflow.wdl test results to Xcelsheet \".tsv\", \".tsv\" or \".xlsx\"\n",
    "    * The python working model of a spreadsheet is the pandas DataFrame.\n",
    "    * This code parses the results directory files content into a DataFrame.\n",
    "    * The options are then to write the DataFrame to a file or display on the command line or both."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "#       Set and verify the Results Directory for this notebook\n",
    "x_dir = '/Users/yo/zzIForge/fullyJan10/'\n",
    "if not os.path.isdir(x_dir):\n",
    "    print('directory not found\\n', x_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_test_results_dataframe(x_dir):\n",
    "    \"\"\" Usage: return_codes_dataframe = get_test_results_dataframe(x_dir)\n",
    "    args:\n",
    "        x_dir:         the directory with the \"call_...\" subdirectories (else you get nothing)\n",
    "        \n",
    "    returns:\n",
    "        rc_df:         pandas dataframe with the return codes and size of various output files\n",
    "    \"\"\"\n",
    "    DATAFRAME_DEFAULT_EMPTY_VALUE = 'unk'\n",
    "    FAILED_RETURN_CODE_READ = '-1'\n",
    "    good_return_codes_list = ['0', '0\\n']\n",
    "    check_files_dict = {'stderr':['ERROR', 'error', 'Error'], 'stdout':['START', 'Finished']}\n",
    "    \n",
    "    #     This variable might be compared to the \"call\" entries in the .wdl tree.\n",
    "    call_dirs = os.listdir(x_dir)\n",
    "    call_dir_list = []\n",
    "    call_dir_count = 0\n",
    "    #     Get the rows list - parse the directories that begin with \"call\" vs getting them from the .wdl files\n",
    "    for call_dir in call_dirs:\n",
    "        if os.path.isdir(os.path.join(call_dir, x_dir)) and call_dir[0:4] == 'call':\n",
    "            call_dir_count += 1\n",
    "            call_dir_list.append(call_dir)\n",
    "    \n",
    "    #     Create the list of things that will be reported in each call directory and initialize the dataframe\n",
    "    cols_list = ['rc', 'bam', 'bam.bai', 'stderr', 'stdout']\n",
    "    rc_df = pd.DataFrame(index=call_dir_list,columns=cols_list).fillna(DATAFRAME_DEFAULT_EMPTY_VALUE)\n",
    "    \n",
    "    #     Check the directories in this tree against the column list\n",
    "    for dir_name, dir_list, files_list in os.walk(x_dir):\n",
    "        for filename in files_list:\n",
    "            full_filename = os.path.join(dir_name, filename)\n",
    "            if filename in cols_list:\n",
    "                if filename == 'rc':\n",
    "                    with open(full_filename, 'r') as fh:\n",
    "                        lines = fh.readlines()\n",
    "                        \n",
    "                    if len(lines) > 0:\n",
    "                        for call_dir in call_dir_list:\n",
    "                            if call_dir in dir_name:\n",
    "                                if lines[0] in good_return_codes_list:\n",
    "                                    rc_df.loc[call_dir, 'rc'] = str(lines[0]).strip()\n",
    "                                else:\n",
    "                                    try:\n",
    "                                        rc_df.loc[call_dir, 'rc'] = str(lines[0]).strip()\n",
    "                                    except:\n",
    "                                        rc_df.loc[call_dir, 'rc'] = FAILED_RETURN_CODE_READ\n",
    "                                        pass\n",
    "                                    \n",
    "            if filename in list(check_files_dict.keys()):\n",
    "                for call_dir in call_dir_list:\n",
    "                    if call_dir in dir_name:\n",
    "                        with open(full_filename, 'r') as fh:\n",
    "                            lines = fh.readlines()\n",
    "\n",
    "                        if len(lines) > 0:\n",
    "                            for line in lines:\n",
    "                                for check_word in check_files_dict[filename]:\n",
    "                                    if check_word in line:\n",
    "                                        rc_df.loc[call_dir, filename] = check_word\n",
    "                                        continue\n",
    "                \n",
    "            # Report the File sizes - code version pending full list of required files\n",
    "            else:\n",
    "                for call_dir in call_dir_list:\n",
    "                    if call_dir in dir_name:\n",
    "                        fname, fext = os.path.splitext(filename)\n",
    "                        this_file_data = os.stat(full_filename)\n",
    "                        if fext[1:] == 'bam':\n",
    "                            rc_df.loc[call_dir, 'bam'] = str(this_file_data.st_size)\n",
    "                        elif fext[1:] == 'bai':\n",
    "                            rc_df.loc[call_dir, 'bam.bai'] = str(this_file_data.st_size)\n",
    "                                                            \n",
    "    return rc_df                 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rc</th>\n",
       "      <th>bam</th>\n",
       "      <th>bam.bai</th>\n",
       "      <th>stderr</th>\n",
       "      <th>stdout</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>call-DHVC</th>\n",
       "      <td>0</td>\n",
       "      <td>unk</td>\n",
       "      <td>unk</td>\n",
       "      <td>unk</td>\n",
       "      <td>unk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>call-realign</th>\n",
       "      <td>0</td>\n",
       "      <td>65349394</td>\n",
       "      <td>1431360</td>\n",
       "      <td>unk</td>\n",
       "      <td>START</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>call-align</th>\n",
       "      <td>0</td>\n",
       "      <td>59500466</td>\n",
       "      <td>1431328</td>\n",
       "      <td>unk</td>\n",
       "      <td>Finished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>call-bqsr</th>\n",
       "      <td>0</td>\n",
       "      <td>67462520</td>\n",
       "      <td>1431360</td>\n",
       "      <td>unk</td>\n",
       "      <td>Finished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>call-dedup</th>\n",
       "      <td>0</td>\n",
       "      <td>65322724</td>\n",
       "      <td>1431360</td>\n",
       "      <td>unk</td>\n",
       "      <td>Finished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>call-haplotype</th>\n",
       "      <td>0</td>\n",
       "      <td>67462520</td>\n",
       "      <td>1431360</td>\n",
       "      <td>unk</td>\n",
       "      <td>Finished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>call-merge</th>\n",
       "      <td>0</td>\n",
       "      <td>59500466</td>\n",
       "      <td>1431328</td>\n",
       "      <td>unk</td>\n",
       "      <td>unk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>call-vqsr</th>\n",
       "      <td>0</td>\n",
       "      <td>unk</td>\n",
       "      <td>unk</td>\n",
       "      <td>unk</td>\n",
       "      <td>Finished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>call-trimseq</th>\n",
       "      <td>0</td>\n",
       "      <td>unk</td>\n",
       "      <td>unk</td>\n",
       "      <td>unk</td>\n",
       "      <td>Finished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>call-DAB</th>\n",
       "      <td>0</td>\n",
       "      <td>65349394</td>\n",
       "      <td>1431360</td>\n",
       "      <td>unk</td>\n",
       "      <td>unk</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               rc       bam  bam.bai stderr    stdout\n",
       "call-DHVC       0       unk      unk    unk       unk\n",
       "call-realign    0  65349394  1431360    unk     START\n",
       "call-align      0  59500466  1431328    unk  Finished\n",
       "call-bqsr       0  67462520  1431360    unk  Finished\n",
       "call-dedup      0  65322724  1431360    unk  Finished\n",
       "call-haplotype  0  67462520  1431360    unk  Finished\n",
       "call-merge      0  59500466  1431328    unk       unk\n",
       "call-vqsr       0       unk      unk    unk  Finished\n",
       "call-trimseq    0       unk      unk    unk  Finished\n",
       "call-DAB        0  65349394  1431360    unk       unk"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "somedf = get_test_results_dataframe(x_dir)\n",
    "somedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7 GermlineMasterWF.DAB.WorkflowJson\n",
      "\t/projects/bioinformatics/DEL/Jsons/GermlineMasterWorkflow.FilledIn.json\n",
      "\n",
      "14 GermlineMasterWF.DHVC.WorkflowJson\n",
      "\t/projects/bioinformatics/DEL/Jsons/GermlineMasterWorkflow.FilledIn.json\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#                                  Be convinced of similar names\n",
    "json_dir = '/Users/yo/zzIForge/Jsons'\n",
    "json_farameters_file = 'GermlineMasterWorkflow.FilledIn.json'\n",
    "import knpackage.toolbox as kn\n",
    "json_parameters = kn.get_run_parameters(json_dir, json_farameters_file)\n",
    "dakine = 'WorkflowJson'\n",
    "lineno = 0\n",
    "for k, v in json_parameters.items():\n",
    "    if dakine in k:\n",
    "        print('%i %30s\\n\\t%s\\n'%(lineno, k, v))\n",
    "    lineno += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
